keps/sig-node/753-sidecar-containers/README.md

```diff
@@ -192,584 +174,1363 @@ logs are lost.
+
+
+#### Beta
+
+- Gather feedback from developers and surveys
+- Complete features A, B, C
+- Additional tests are in Testgrid and linked in KEP
+
+#### GA
+
+- N examples of real-world usage
+- N installs
+- More rigorous forms of testing—e.g., downgrade tests and scalability tests
+- Allowing time for feedback
+
+**Note:** Generally we also wait at least two releases between beta and
+GA/stable, because there's no opportunity for user feedback, or even bug reports,
+in back-to-back releases.
+
+**For non-optional features moving to GA, the graduation criteria must include
+[conformance tests].**
+
+[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md
+
+#### Deprecation
+
+- Announce deprecation and support policy of the existing flag
+- Two versions passed since introducing the functionality that deprecates the flag (to address version skew)
+- Address feedback on usage/changed behavior, provided on GitHub issues
+- Deprecate the flag
+-->
 
 ### Upgrade / Downgrade Strategy
-When upgrading no changes should be needed to maintain existing behaviour as all of this behaviour is optional and disabled by default.
-To activate the feature they will need to enable the feature gate and mark their containers as sidecars in the container spec.
 
-When downgrading `kubectl`, users will need to remove the sidecar field from any of their Kubernetes manifest files as `kubectl` will refuse to apply manifests with an unknown field (unless you use `--validate=false`).
+#### Upgrade strategy
+
+Existing sidecars (implemented as regular containers) will still work as
+intended, as in the past we don't recognize the new field `restartPolicy` today.
+
+Upgrade will not change any current behaviors.
+
+#### Downgrade strategy
+
+First, there will be no effect on any workload that doesn't use a new field. Any
+combination of feature gate enabled/disabled or version skew will work as usual
+for that workload.
+
+So when the new functionality wasn't yet used, downgrade will not be affected.
+
+Due to the new field added to `initContainers` to turn them into sidecars,
+downgrading to the version without this feature will make all Pods using this
+flag unscheduleable. These Pod spec will be rejected by the control plane and
+all kubelets during unmarshalling.
+
+**Note**, this specific behavior on read is not clear and if unmarshaling works,
+the new logic to reject such Pods will be added.
+See [Upgrade/downgrade testing](#upgradedowngrade-testing) section.
+
+Workloads will have to be deleted and recreated with the old way of handling
+sidecars.  Once there is no more Pods using sidecars, node can be downgraded
+without side effects.
+
+If downgrade hapenning from the version with the feature enabled to the previous
+version that has this feature support, but feature gate is disabled, kubelet
+and/or control place will reject these Pods.
```

Comment:
Regarding control-plane - what we're doing at the API level is we're (silently) dropping the field that has been newly set. However, if the field was already set, we keep that set.
[The note below explains why this is fine - just maybe clarify it here.]

Reply:
Updated the section "#### Downgrade strategy".